### PR TITLE
Magical Steps in Auglag

### PR DESCRIPTION
**Pull request by lambe**
A fast implementation of the "magical steps" for Auglag to exploit problem structure and improve performance.

This update also includes a couple of small tweaks to Auglag in the first commit.

### CODE REVIEW

**Comment by dpo** [comment]
Thanks. This looks good. Could you fix the `.travis.yml` so the CI build passes though?

**Comment by lambe** [comment]
Note that I have added a couple of additional commits. Please review again.

File: nlp/optimize/auglag.py [hunk 1]
```diff
@@ -182,12 +182,12 @@ def least_squares_multipliers(self, x):
...
                                free_vars)
 
-        g = slack_model.grad(x)
+        g = slack_model.grad(x) - J.T * al_model.pi
 
         lsqr = LSQRSolver(Jred.T)
         lsqr.solve(g[free_vars], itnlim=lim)
         if lsqr.optimal:
-            al_model.pi = lsqr.x.copy()
+            al_model.pi += lsqr.x.copy()
```
Comment (dpo) [inline nlp/optimize/auglag.py]: Is this behavior better numerically?

Comment (lambe) [inline nlp/optimize/auglag.py]: Yes. A couple of the CUTEst problems had difficulty converging when calculating the multipliers from scratch close to the optimal solution. This was fixed by computing the change in multipliers instead.

**Comment by dpo** [comment]
Try `brew uninstall -f` so `brew` doesn't error out when attempting to uninstall a package that isn't installed.

**Comment by lambe** [comment]
Now we're back to the build error I had before. It has something to do with a dependency conflict in Homebrew.

**Comment by dpo** [comment]
I think you have to leave `--ignore-dependencies` and add `-f`.

**Pull request merged**